Floating Point Error Estimation Framework in Clad (#247)
vgvassilev/clad: demos/ErrorEstimation/CustomModel/README.md

> # Using your custom estimation model with clad 

Reviewer: @davidlange6, can you proofread that section?

> Clad's error estimation framework provides users with a choice to write their own custom error models and use those to generate estimation code. By default clad uses the Taylor series approximation model to estimate errors in a function, this model however might not be the most sutiable alternative in many cases. In such scenarios, a user might prefer writing their own models. 

Reviewer: ```suggestion
Clad's floating point (fp) error estimation framework provides users with a choice to write their own custom models and use those to generate fp estimation code. By default clad uses the Taylor series approximation model to estimate fp errors in a function, this model however might not be the most suitable alternative in many cases. In such scenarios, a user might prefer writing their own models. 
```

> Clad's floating point (fp) error estimation framework provides users with a choice to write their own custom models and use those to generate fp estimation code. By default clad uses the Taylor series approximation model to estimate fp errors in a function, this model however might not be the most suitable alternative in many cases. In such scenarios, a user might prefer writing their own models. 

Reviewer 2: I think "choice" is the wrong word (without knowing any details..) - how about "provides users the capability to generate fp estimation code based on their custom models."

add "however" before "this model" 
and "may prefer to write their own model to get better fp error estimates."

Author: > I think "choice" is the wrong word (without knowing any details..) - how about "provides users the capability to generate fp estimation code based on their custom models."

The reason for using 'choice' there is that users may write their own custom model OR may use the one provided by clad. So I feel like 'choice' is a better word there? Not sure...

> The aim of this demo is to illustrate how one can integrate thier custom model with clad. For in depth information on how to *write* your own custom estimation model, check out the tutorials here.

Reviewer: ```suggestion
The aim of this demo is to illustrate how one can integrate their custom model with clad. In-depth information on how to *write* a custom fp estimation model can be found in the tutorials here.
```

> The aim of this demo is to illustrate how one can integrate their custom model with clad. In-depth information on how to *write* a custom fp estimation model can be found in the tutorials here.

Reviewer 2: their->a

 <!--- TODO: Add doc link -->

## Building the demo

> Before we can use the custom model, we should compile it into a [shared object](https://www.thegeekstuff.com/2012/06/linux-shared-libraries/). To do this, you can use your favourite compiler. For this demo, we will be using the clang compiler.

Reviewer 2: it must be compiled into ..

In a terminal, run the following:

```bash
$ /path/to/clang -Ipath/to/install/include -fPIC -shared -fno-rtti -Wl,-undefined -Wl,suppress path/to/CustomModel.cpp -o libCustomModel.so
``` 
 The above should create a ```libCustomModel.so``` in the same directory you executed that command in. Once the shared object is created, we are ready to run it with clad.

## Running the demo

> Now, to use your custom estimation model, you can just specify the ```.so``` created in the previous section as an input to clad via CLI. The specific parameters you would need to add are given below:

Reviewer: ```suggestion
Now, to use your custom fp estimation model, you can just specify the ```.so``` created in the previous section as an input to clad via the command line. The specific parameters you would need to add are given below:
```

```bash
-Xclang -plugin-arg-clad -Xclang -fcustom-estimation-model -Xclang -plugin-arg-clad -Xclang ./libCustomModel.so
``` 
So a typical invocation to clad would then look like the following:

```bash

> ./bin/clang -Xclang -add-plugin -Xclang clad -Xclang -load -Xclang path/to/clad.so -I../include/ -x c++ -lstdc++ -Xclang -plugin-arg-clad -Xclang -fcustom-estimation-model -Xclang -plugin-arg-clad -Xclang ./libCustomModel.so /path/to/some.cpp

Reviewer 2: I would suggest to use commands that actually work in the demo (and to explain under what circumstances they work. Eg,  I'd suggest

clang -Xclang..... ${CLAD_BASE}/lib/clad.so ... lib/libCustomModel.so src/model.cpp

(or can clad.so be found by the usual -L mechanism instead?)

Reviewer 2: eg, the names can also be changed to correspond to a test case?

Author: > eg, the names can also be changed to correspond to a test case?

Which names exactly? ```CustomModel``` used there is actually the name of the file in the demo too.

> (or can clad.so be found by the usual -L mechanism instead?)

from what I know, the lib-name has to be prefixed with a ```lib``` (for eg. ```libclad.so```) for that to work.

> I would suggest to use commands that actually work in the demo (and to explain under what circumstances they work. Eg, I'd suggest
>
> clang -Xclang..... ${CLAD_BASE}/lib/clad.so ... lib/libCustomModel.so src/model.cpp

But for that, I would need to specify what ```CLAD_BASE``` points to right? Which will have to be some dummy value again like ```path/to/clad/installation``` or something?

Reviewer 2: I'd suggest assuming (and stating) that clang is in ones PATH. so just "clang" not bin/clang or /path/to/clang

Then yes, you would specify how to set CLAD_BASE. eg
git clone git@github.com:clad/clad
setenv CLAD_BASE $PWD/clad
(or its bash equivalent)

```bash
./bin/clang -Xclang -add-plugin -Xclang clad -Xclang -load -Xclang path/to/clad.so -I../include/ -x c++ -lstdc++ -Xclang -plugin-arg-clad -Xclang -fcustom-estimation-model -Xclang -plugin-arg-clad -Xclang ./libCustomModel.so /path/to/some.cpp
```
## Verifying results

To verify your results, you can build the dummy ```test.cpp``` file with the commands shown above. Once you compile and run the test file correctly, you will notice the generated code is as follows:

```cpp
The code is: void func_grad(float x, float y, clad::array_ref<float> _d_x, clad::array_ref<float> _d_y, double &_final_error) {
    double _delta_z = 0;
    float _EERepl_z0;
    float _d_z = 0;
    float _EERepl_z1;
    float z;
    _EERepl_z0 = z;
    z = x + y;
    _EERepl_z1 = z;
    float func_return = z;
    goto _label0;
  _label0:
    _d_z += 1;
    {
        float _r_d0 = _d_z;
        * _d_x += _r_d0;
        * _d_y += _r_d0;
        _delta_z += _r_d0 * _EERepl_z1;
        _d_z -= _r_d0;
    }
    double _delta_x = 0;
    _delta_x += * _d_x * x;
    double _delta_y = 0;
    _delta_y += * _d_y * y;
    _final_error += _delta_y + _delta_z + _delta_x;
}

> ```

Reviewer: Perhaps reword this section taking into account that we have a runnable demo.
